new/rfcXXX-policyfile-endpoint.md

```diff
@@ -0,0 +1,482 @@
+converge a node (host) to a desired state. In contrast with the current
+behaviors of Chef Client and Server, Policyfiles provide a way to
+describe and manage configuration code on a whole-node basis, reducing
+Chef Client's use of shared, mutable resources hosted on the server and
+providing operators easier visibility into configuration code and
+changes. For further description of the Policyfile feature, read the dedicated
+[POLICYFILE README.](https://github.com/chef/chef-dk/blob/master/POLICYFILE_README.md)
+
+A preview implementation of the Policyfile feature is currently available; in
+this implementation, Policyfile documents are stored as data bag items.
+Although this implementation works, it does not provide the baseline quality
+users expect. In particular, there is no validation, no means to make any
+relational queries, and documents are indexed for search as data bag items.
+
+This RFC describes a new Policyfile HTTP Resource API that remediates these
+issues.
+
+## Motivation
+
+    As a Chef user,
+    I want Policyfiles to have a dedicated HTTP Resource API,
+    so that Policyfiles will be validated upon upload
+    and I can query the relationships between Policyfiles and other objects,
+    and Policyfiles will not be indexed as an incorrect data type.
+
+## Specification
+
+### General Concepts
+
+A Policyfile lock is a JSON document that specifies a node's run list and the
+set of cookbooks a node will use to converge the node.
+
+A Policy Group is a set of nodes defined by a single token. Nodes must be a
+member of a Policy Group in order to use the Policyfile feature, and nodes may
+only be a member of one policy group.
+
+### Policyfile Lock
+
+A Policyfile lock is a JSON document that has the following structure:
+
+```json
+{
+  "revision_id": "edd40c30c4e0ebb3658abde4620597597d2e9c17",
+  "name": "some_policy_name",
+  "run_list": [
+    "recipe[policyfile_example_cookbook::default]"
+  ],
+  "cookbook_locks": {
+    "policyfile_example_cookbook": {
+      "version": "1.0.0",
+      "identifier": "f04cc40faf628253fe7d9566d66a1733fb1afbe9"
+    }
+  }
+}
+
+```
+
+#### `revision_id` Field (required)
+
+The `revision_id` field is used to distinguish revisions of a Policyfile lock
+document. It is a JSON String. It may contain alphanumeric characters, hyphens,
+underscores, the dot character and the colon character. In regular expression
+form: `/^[\-[:alnum:]_\.\:]+$/`. The revision ID is limited to a maximum of 255
+characters. The Chef Server must reject a Policyfile lock document when the
+revision ID does not meet these requirements.
+
+In ChefDK the revision will be calculated by hashing other content in the
+Policyfile lock document and thus will be a hexidecimal number. In order to
+allow alternate implementations to generate revision IDs according to different
+schemes (perhaps timestamps or Ci build numbers), this field will not be
+required to be a hexidecimal string.
```

Comment:
I think the revision id should be a strict cryptographic hash of it's canonicalized contents, and that should be enforced at the server. Allow a metadata field for user defined things like timestamps/CI build numbers, but don't conflate the revision_id with the metadata. The lockfile goes to some effort refer to the exact content of the cookbook (the identifier) and moves away from using the version to ; let's carry that all the way through into the lockfile id scheme as well. 

Otherwise you could create a policyfile lock with some collection of cookbooks, and replace it with a different one with a different collection of cookbooks, defeating much of the benefit of moving away from version numbers. This is especially important because the policy groups seem to essentially refer to lockfiles by reference to that id. 

Having the policy groups include policy lockfiles by reference to the id, and allowing the id to be arbitrary creates some possible issues around permissions. Anything short of requiring the policy lockfile revision id to be a strict hash of its contents, and enforcing that at the server level is going to create nasty opportunities. For example dev creates policy id foobar, which passes acceptance, but deletes/recreates it with a policy with the same name but some sneaky property changes in between acceptance and prod deploy.